src/ota/mod.rs

```diff
@@ -104,3 +126,2108 @@ impl Default for DeployStatus {
+            Ok(state) => state,
+            Err(err) => {
+                let message = "Unable to read pending ota state".to_string();
+                error!("{message} : {}", err);
+                return OtaStatus::Failure(OtaError::Io(message), None);
+            }
+        };
+
+        let request_uuid = ota_state.uuid;
+        let ota_request = OtaId {
+            uuid: request_uuid,
+            url: "".to_string(),
+        };
+
+        if let Err(error) = self.do_pending_ota(&ota_state).await {
+            return OtaStatus::Failure(error, Some(ota_request));
+        }
+
+        OtaStatus::Success(ota_request)
+    }
+
+    pub async fn do_pending_ota(&self, state: &PersistentState) -> Result<(), OtaError> {
+        const GOOD_STATE: &str = "good";
+
+        let booted_slot = self.system_update.boot_slot().await.map_err(|error| {
+            let message = "Unable to identify the booted slot";
+            error!("{message}: {error}");
+            OtaError::Internal(message)
+        })?;
+
+        if state.slot == booted_slot {
+            let message = "Unable to switch slot";
+            return Err(OtaError::SystemRollback(message));
+        }
+
+        let primary_slot = self.system_update.get_primary().await.map_err(|error| {
+            let message = "Unable to get the current primary slot";
+            error!("{message}: {error}");
+            OtaError::Internal(message)
+        })?;
+
+        let (marked_slot, _) = self
+            .system_update
+            .mark(GOOD_STATE, &primary_slot)
+            .await
+            .map_err(|error| {
+                let message = "Unable to run marking slot operation";
+                error!("{message}: {error}");
+                OtaError::Internal(message)
+            })?;
+
+        if primary_slot != marked_slot {
+            let message = "Unable to mark slot";
+            Err(OtaError::Internal(message))
+        } else {
+            Ok(())
+        }
+    }
+
+    pub async fn next(&mut self) {
+        self.ota_status = match self.ota_status.clone() {
+            OtaStatus::Init(req) => OtaStatus::Acknowledged(req),
+            OtaStatus::Acknowledged(ota_request) => OtaStatus::Downloading(ota_request, 0),
+            OtaStatus::Downloading(ota_request, _) => self.download(ota_request).await,
+            OtaStatus::Deploying(ota_request, _) => self.deploy(ota_request).await,
+            OtaStatus::Deployed(ota_request) => self.reboot(ota_request).await,
+            OtaStatus::Rebooted => self.check_reboot().await,
+            OtaStatus::Error(ota_error, ota_request) => {
+                OtaStatus::Failure(ota_error, Some(ota_request))
+            }
+            rebooting @ OtaStatus::Rebooting(_) => rebooting,
+            OtaStatus::NoPendingOta | OtaStatus::Success(_) | OtaStatus::Failure(_, _) => {
+                OtaStatus::Idle
+            }
+            OtaStatus::Idle => OtaStatus::Idle,
```

Comment:
Is this put separately from the previous to emphasize that, if the state is Idle, it will remain Idle once `next` is called? Otherwise you could merge it with the previous ones

Reply:
Yes, it was to make it clearer, but it's the same anyway. Joined with the previous pattern